lib/SILGen/ConcurrencyUtils.cpp

```diff
@@ -0,0 +1,163 @@
+                                               LookUpConformanceInModule()),
+                          {metatypeVal});
+  auto one = SGF.B.createIntegerLiteral(loc, silWordType, 1);
+  auto three = SGF.B.createIntegerLiteral(loc, silWordType, 3);
+  auto four = SGF.B.createIntegerLiteral(loc, silWordType, 4);
+  auto valueToShiftLit =
+      SGF.B.createIntegerLiteral(loc, silWordType, valueToShift);
+
+  // sizeof(Word) - 1
+  auto sub = SGF.B.createBuiltinBinaryFunction(loc, "sub", silWordType,
+                                               silWordType, {sizeOfWord, one});
+  // (sizeof(Word) - 1) << 3
+  auto innerShift = SGF.B.createBuiltinBinaryFunction(
+      loc, "shl", silWordType, silWordType, {sub, three});
+  // ((sizeof(Word) - 1) << 3) + 4
+  auto innerShiftOffset = SGF.B.createBuiltinBinaryFunction(
+      loc, "add", silWordType, silWordType, {innerShift, four});
+  auto outerShift =
+      SGF.B.createBuiltinBinaryFunction(loc, "shl", silWordType, silWordType,
+                                        {valueToShiftLit, innerShiftOffset});
+  return outerShift;
+}
+
+/// Construct the TBI mask in a platform independent way that works on all
+/// platforms.
+///
+/// We compute:
+///
+///   mask = (0x3 << (((sizeof(Word) - 1) << 3) + 4)) ^ -1
+static SILValue getTBIClearMask(SILGenFunction &SGF, SILLocation loc) {
+  auto &ctx = SGF.getASTContext();
+  auto silWordType = SILType::getBuiltinWordType(ctx);
+  auto negBits = SGF.B.createIntegerLiteral(loc, silWordType, -1);
+
+  return SGF.B.createBuiltinBinaryFunction(loc, "xor", silWordType, silWordType,
+                                           {getTBIBits(SGF, loc, 3), negBits});
+}
+
+static ManagedValue
+transformTupleElts(SILGenFunction &SGF, SILLocation loc, ManagedValue mv,
+                   llvm::function_ref<SILValue(DestructureTupleInst *)> func) {
+  auto &ctx = SGF.getASTContext();
+  auto silWordType = SILType::getBuiltinWordType(ctx);
+  auto tupleType =
+      SILType::getTupleType(SGF.getASTContext(), {silWordType, silWordType});
+  auto cast =
+      SGF.B.createUncheckedBitCast(loc, mv, tupleType).getUnmanagedValue();
+  auto *destructure = SGF.B.createDestructureTuple(loc, cast);
+  SILValue reformedValue = func(destructure);
+  auto reformedPointer =
+      ManagedValue::forBorrowedRValue(SGF.B.createUncheckedOwnershipConversion(
+          loc,
+          SGF.B.createUncheckedReinterpretCast(loc, reformedValue,
+                                               mv.getType()),
+          OwnershipKind::Guaranteed));
+  return SGF.B.createMarkDependence(loc, reformedPointer, mv,
+                                    MarkDependenceKind::NonEscaping);
+}
+
+ManagedValue swift::Lowering::clearImplicitIsolatedActorBits(
+    SILGenFunction &SGF, SILLocation loc, ManagedValue isolatedMV) {
+  auto &ctx = SGF.getASTContext();
+  if (!ctx.LangOpts.hasFeature(Feature::NonisolatedNonsendingDynamicHopElim) &&
+      !ctx.LangOpts.HasAArch64TBI) {
```

Comment:
Why are we checking for TBI here? It seems like we effectively need to distinguish between places where TBI will strip the bits for us (e.g., we're dereferencing the pointer) vs those where we treat the pointer as a value and we need to strip the bits ourselves (e.g., when building a runtime data structure involving this witness table).